File: src/std/js/ConcurrentMap.js

fan.std.ConcurrentMap = fan.sys.Obj.$extend(fan.sys.Obj);
fan.std.ConcurrentMap.prototype.$ctor = function() {}
fan.std.ConcurrentMap.prototype.$typeof = function() { return fan.sys.AtomicRef.$type; }

fan.std.ConcurrentMap.make = function(capacity) {
  self = new fan.std.ConcurrentMap();
  if (capacity === undefined) v = 256;
  self.m_val = new Map();
  return self;
}

fan.std.ConcurrentMap.prototype.isEmpty = function() {
  return this.m_val.size == 0;
}

fan.std.ConcurrentMap.prototype.size = function() {
  return this.m_val;
}

fan.std.ConcurrentMap.prototype.get = function(k) {
  return this.m_val.get(k);
}

fan.std.ConcurrentMap.prototype.set = function(k, v) {
  fan.std.ConcurrentMap.checkImmutable(v);
  return this.m_val.set(k, v);
}

fan.std.ConcurrentMap.prototype.add = function(k, v) {
  if (this.m_val.has(k)) throw fan.sys.Err.make("Duplicate key:  " + key);
  fan.std.ConcurrentMap.checkImmutable(values);
  return this.m_val.set(k, v);
}

fan.std.ConcurrentMap.prototype.getOrAdd = function(k, defVal) {
  if (this.m_val.has(k)) {
    return this.m_val.get(k)
  }
  this.m_val.set(k, defVal)
  return defVal;
}

fan.std.ConcurrentMap.prototype.setAll = function(m) {
  for (var [key, value] of this.m_val) {
    fan.std.ConcurrentMap.checkImmutable(value);
    m.set(key, value)
  }
  return this;
}

fan.std.ConcurrentMap.prototype.remove = function(k) {
  old = this.m_val.get(k)
  this.m_val.delete(k)
  return old;
}

fan.std.ConcurrentMap.prototype.clear = function(m) {
  this.m_val.clear();
}


fan.std.ConcurrentMap.prototype.each = function(f) {
  for (var [key, value] of this.m_val) {
    f.call(value, key)
  }
}

fan.std.ConcurrentMap.prototype.eachWhile = function(f) {
  for (var [key, value] of this.m_val) {
    var r = f.call(value, key)
    if (r == null) return r;
  }
  return null;
}

fan.std.ConcurrentMap.prototype.containsKey = function(k) {
  return this.m_val.has(k)
}

fan.std.ConcurrentMap.prototype.keys = function() {
  var list = fan.sys.List.make(this.m_val.size);
  for (var key of myMap.keys()) {
    list.add(key)
  }
  return list;
}

fan.std.ConcurrentMap.prototype.vals = function() {
  var list = fan.sys.List.make(this.m_val.size);
  for (var key of myMap.values()) {
    list.add(key)
  }
  return list;
}

fan.std.ConcurrentMap.checkImmutable = function(val) {
  if (fan.sys.Obj.isImmutable(val))
    return val;
  else
    throw NotImmutableErr.make();
}

File: library/std/java/src/fan/std/ConcurrentMap.java
//
// Copyright (c) 2016, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   11 Feb 16  Brian Frank  Creation
//

package fan.std;

import java.util.Iterator;
import java.util.Map.Entry;
import java.util.concurrent.ConcurrentHashMap;
import fan.sys.*;
import fanx.interop.Interop;
import fanx.main.Type;

/**
 * ConcurrentMap
 */
public final class ConcurrentMap extends FanObj
{
  public static ConcurrentMap make() { return make(256); }
  public static ConcurrentMap make(long capacity) { return new ConcurrentMap((int)capacity); }

  ConcurrentMap(int capacity) { this.map = new ConcurrentHashMap(capacity); }

  public final Type typeof() { return typeof; }
  private static final Type typeof = Type.find("std::ConcurrentMap");

  public boolean isEmpty() { return map.size() == 0; }

  public long size() { return map.size(); }

  public Object get(Object key) { return map.get(key); }

  public void set(Object key, Object val) { map.put(key, checkImmutable(val)); }

  public Object getAndSet(Object key, Object val)
  {
    return map.put(key, checkImmutable(val));
  }
  
  public void add(Object key, Object val) {
    if (map.putIfAbsent(key, checkImmutable(val)) != null)
      throw Err.make("Duplicate key:  " + key);
  }

  public Object getOrAdd(Object key, Object defVal)
  {
    Object val = map.putIfAbsent(key, checkImmutable(defVal));
    return val == null ? defVal : val;
  }

  public ConcurrentMap setAll(Map m)
  {
    if (m.isImmutable()) map.putAll(Interop.toJava(m.rw()));
    else
    {
      final List vals = m.vals();
      for (int i=0; i<vals.size(); ++i) checkImmutable(vals.get(i));
      map.putAll(Interop.toJava(m));
    }
    return this;
  }

  public Object remove(Object key) { return map.remove(key); }

  public void clear() { map.clear(); }

  public void each(Func f)
  {
    Iterator it = map.entrySet().iterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      f.call(e.getValue(), e.getKey());
    }
  }

  public Object eachWhile(Func f)
  {
    Iterator it = map.entrySet().iterator();
    while (it.hasNext())
    {
      Entry e = (Entry)it.next();
      Object r = f.call(e.getValue(), e.getKey());
      if (r != null) return r;
    }
    return null;
  }

  public boolean containsKey(Object key) { return map.containsKey(key); }

  public List keys()
  {
    List list = List.make(map.size());
    Iterator it = map.keySet().iterator();
    while (it.hasNext()) list.add(it.next());
    return list;
  }

  public List vals()
  {
    List list = List.make(map.size());
    Iterator it = map.values().iterator();
    while (it.hasNext()) list.add(it.next());
    return list;
  }

  private Object checkImmutable(Object val)
  {
    if (FanObj.isImmutable(val))
      return val;
    else
      throw NotImmutableErr.make();
  }

  final ConcurrentHashMap map;
}